File: INFORMÁTICA/PROGRAMACIÓN Y DISEÑO ORIENTADO A OBJETOS/PRÁCTICAS/PROYECTOS/EXÁMENES/examen_parcial_20-21/Deepspace/BetaPowerEfficientSpaceStation.rb
#encoding: UTF-8

require_relative 'PowerEfficientSpaceStation'
require_relative 'BetaPowerEfficientSpaceStationToUI'
require_relative 'Dice'

module Deepspace

    #Encapsula la transformacion a BetaPowerEfficientSpaceStation
    class BetaPowerEfficientSpaceStation < PowerEfficientSpaceStation

        
        #Potencia extra de disparo y proteccion.
        @@EXTRAEFFICIENCY=1.2

        #param: station --> SpaceStation
        def initialize (station)
            super(station)

            @dice = Dice.new
        end

        #Interfaz

        def getUIversion
            return (BetaPowerEfficientSpaceStationToUI.new(self))
        end

        def fire()
            fire=super
            if(@dice.extraEfficiency()) 
                fire*=@@EXTRAEFFICIENCY
            end

            return fire
        end


    end
end

File: INFORMÁTICA/PROGRAMACIÓN Y DISEÑO ORIENTADO A OBJETOS/PRÁCTICAS/PROYECTOS/EXÁMENES/examen_2_19-20/deepspace/src/deepspace/BetaPowerEfficientSpaceStation.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

/**
 *
 * @author leandrofdez
 */

//Encapsula la transformacion a BetaPowerEfficientSpaceStation. 
public class BetaPowerEfficientSpaceStation extends PowerEfficientSpaceStation{
    
    
    //Potencia extra de disparo.
    private static final float EXTRAEFFICIENCY=1.2f;
    
    //Dado que nos permitira determinar si se dispara con mas potencia.
    Dice dice;
    
    BetaPowerEfficientSpaceStation(SpaceStation station){
        super(station);
        dice=new Dice();
        
    }
    
    
    /*
    Añade un factor extra a la potencia disparo.
    */
    @Override
    public float fire(){
        
        float fire=super.fire();
        if (dice.extraEfficiency()) fire*=EXTRAEFFICIENCY;
        
        return fire;
    }
    
    //Interfaz
    @Override
    public BetaPowerEfficientSpaceStationToUI getUIversion(){
        return (new BetaPowerEfficientSpaceStationToUI(this));
    }
    
}
